apollo-router-core/src/services/http_compat.rs

```diff
@@ -15,82 +16,63 @@ use bytes::Bytes;
 #[cfg(feature = "axum-server")]
 use crate::ResponseBody;
 
-use http::{
-    header::HeaderName,
-    request::{Builder, Parts},
-    uri::InvalidUri,
-    HeaderMap, HeaderValue, Method, Uri, Version,
-};
+use http::{header::HeaderName, request::Parts, uri::InvalidUri, HeaderValue, Method, Uri};
 
 #[derive(Debug)]
 pub struct Request<T> {
-    // The goal of having a copy of the url is to keep the right type for `ReqwestSubgraphService` and avoid re-parsing.
-    // This url will stay the same than the uri in inner because we only can set a new url with `set_url` method
-    pub(super) url: Uri,
     inner: http::Request<T>,
 }
 
+// Most of the required functionality is provided by our Deref and DerefMut implementations.
+#[buildstructor::builder]
 impl<T> Request<T> {
+    /// This is the constructor (or builder) to use when constructing a real Request.
+    ///
+    /// Required parameters are required in non-testing code to create a Request.
+    pub fn new(
+        headers: Vec<(HeaderName, HeaderValue)>,
+        uri: http::Uri,
+        method: http::Method,
+        body: T,
+    ) -> http::Result<Request<T>> {
+        let mut builder = http::request::Builder::new().method(method).uri(uri);
+        for (key, value) in headers {
+            builder = builder.header(key, value);
+        }
+        let req = builder.body(body)?;
+
+        Ok(Self { inner: req })
+    }
+
+    /// This is the constructor (or builder) to use when constructing a "fake" Request.
+    ///
+    /// This does not enforce the provision of the uri and method that is required for a fully functional
+    /// Request. It's usually enough for testing, when a fully consructed Request is
+    /// difficult to construct and not required for the purposes of the test.
+    pub fn fake_new(
+        headers: Vec<(HeaderName, HeaderValue)>,
```

Comment:
The fake should probably take headers as `Vec<(String, String)>`
The conversion can happen in the constructor as it is fallible.

Reply:
This was something that @bnjjj introduced. I originally had a `Vec<(String, String)>` or something like that and he suggested this change because it's possible to have multiple values for the same header. I don't think that prohibits the use of (String, String), so I may not be explaining it very well in my befuddled state. @bnjjj Can you discuss this with @BrynCooke  and decide how you want to proceed? For now, I'd prefer not to do more changes with this PR and defer any more work to a follow up PR, but I'm follow up with whatever you guys decide.